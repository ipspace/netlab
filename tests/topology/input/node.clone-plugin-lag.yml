plugin: [node.clone]

module: [lag, vlan]

vlans:
  red:

groups:
  _auto_create: true
  routers:
    members: [r1, r2]      # Devices that supports (m)lag
    device: eos
    module: [vlan, lag]

nodes:
  h:                       # Short name, regular test
    device: frr
    clone.count: 2

  h2:
    device: frr
    clone.count: 2

links:
- lag:
    members: [r1-r2]
    mlag.peergroup: true
- lag:
    members: [r1-h, r1-h]
    ifindex: 8             # Custom lag.ifindex, gets incremented for each clone
  pool: l2only
- lag:
    members:
    - r2:
        ifindex: 6         # Custom ifindex on interface, gets incremented for each clone
      h:
    - r2:
        ifindex: 7
      h:
  pool: l2only
- lag:
    members: [h2-r1, h2-r2]  # mlag, avoid issue with overlapping lag.ifindex
  vlan.access: red
