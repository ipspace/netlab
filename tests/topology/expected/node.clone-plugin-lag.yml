groups:
  routers:
    device: eos
    members:
    - r1
    - r2
    module:
    - vlan
    - lag
input:
- topology/input/node.clone-plugin-lag.yml
- package:topology-defaults.yml
lag:
  lacp: fast
  lacp_mode: active
links:
- _linkname: links[1].peerlink[1]
  interfaces:
  - ifindex: 1
    ifname: Ethernet1
    node: r1
  - ifindex: 1
    ifname: Ethernet1
    node: r2
  lag:
    mlag:
      peergroup: 1
  linkindex: 1
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[1]
  interfaces:
  - ifindex: 2
    ifname: Ethernet2
    node: r1
  - ifindex: 1
    ifname: eth1
    node: h-01
  lag:
    _parentindex: 2
  linkindex: 8
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[2].lag[2]
  interfaces:
  - ifindex: 3
    ifname: Ethernet3
    node: r1
  - ifindex: 2
    ifname: eth2
    node: h-01
  lag:
    _parentindex: 2
  linkindex: 9
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[3].lag[1]
  interfaces:
  - ifindex: 4
    ifname: Ethernet4
    node: r1
  - ifindex: 1
    ifname: eth1
    node: h-02
  lag:
    _parentindex: 3
  linkindex: 10
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[3].lag[2]
  interfaces:
  - ifindex: 5
    ifname: Ethernet5
    node: r1
  - ifindex: 2
    ifname: eth2
    node: h-02
  lag:
    _parentindex: 3
  linkindex: 11
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[1]
  interfaces:
  - ifindex: 6
    ifname: Ethernet6
    node: r2
  - ifindex: 3
    ifname: eth3
    node: h-01
  lag:
    _parentindex: 4
  linkindex: 12
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[4].lag[2]
  interfaces:
  - ifindex: 7
    ifname: Ethernet7
    node: r2
  - ifindex: 4
    ifname: eth4
    node: h-01
  lag:
    _parentindex: 4
  linkindex: 13
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[5].lag[1]
  interfaces:
  - ifindex: 8
    ifname: Ethernet8
    node: r2
  - ifindex: 3
    ifname: eth3
    node: h-02
  lag:
    _parentindex: 5
  linkindex: 14
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[5].lag[2]
  interfaces:
  - ifindex: 9
    ifname: Ethernet9
    node: r2
  - ifindex: 4
    ifname: eth4
    node: h-02
  lag:
    _parentindex: 5
  linkindex: 15
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[6].lag[1]
  interfaces:
  - ifindex: 1
    ifname: eth1
    node: h2-01
  - ifindex: 6
    ifname: Ethernet6
    node: r1
  lag:
    _parentindex: 6
  linkindex: 16
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[6].lag[2]
  interfaces:
  - ifindex: 2
    ifname: eth2
    node: h2-01
  - ifindex: 2
    ifname: Ethernet2
    node: r2
  lag:
    _parentindex: 6
  linkindex: 17
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[7].lag[1]
  interfaces:
  - ifindex: 1
    ifname: eth1
    node: h2-02
  - ifindex: 7
    ifname: Ethernet7
    node: r1
  lag:
    _parentindex: 7
  linkindex: 18
  node_count: 2
  prefix: false
  type: p2p
- _linkname: links[7].lag[2]
  interfaces:
  - ifindex: 2
    ifname: eth2
    node: h2-02
  - ifindex: 3
    ifname: Ethernet3
    node: r2
  lag:
    _parentindex: 7
  linkindex: 19
  node_count: 2
  prefix: false
  type: p2p
module:
- lag
- vlan
name: input
nodes:
  h-01:
    af:
      ipv4: true
    box: debian/bookworm64
    device: frr
    id: 3
    interfaces:
    - ifindex: 30000
      ifname: bond8
      lag:
        ifindex: 8
        lacp: fast
        lacp_mode: active
      name: h-01 -> r1
      neighbors:
      - ifname: port-channel8
        node: r1
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30001
      ifname: bond9
      lag:
        ifindex: 9
        lacp: fast
        lacp_mode: active
      name: h-01 -> r2
      neighbors:
      - ifname: port-channel1
        node: r2
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 8
      linkindex: 8
      name: h-01 -> r1
      neighbors:
      - ifname: Ethernet2
        node: r1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 8
      linkindex: 9
      name: h-01 -> r1
      neighbors:
      - ifname: Ethernet3
        node: r1
      type: p2p
    - ifindex: 3
      ifname: eth3
      lag:
        _parentindex: 9
      linkindex: 12
      name: h-01 -> r2
      neighbors:
      - ifname: Ethernet6
        node: r2
      type: p2p
    - ifindex: 4
      ifname: eth4
      lag:
        _parentindex: 9
      linkindex: 13
      name: h-01 -> r2
      neighbors:
      - ifname: Ethernet7
        node: r2
      type: p2p
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.3/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.103
      mac: 08:4f:a9:00:00:03
    module:
    - lag
    - vlan
    name: h-01
  h-02:
    af:
      ipv4: true
    box: debian/bookworm64
    device: frr
    id: 4
    interfaces:
    - ifindex: 30000
      ifname: bond9
      lag:
        ifindex: 9
        lacp: fast
        lacp_mode: active
      name: h-02 -> r1
      neighbors:
      - ifname: port-channel9
        node: r1
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30001
      ifname: bond10
      lag:
        ifindex: 10
        lacp: fast
        lacp_mode: active
      name: h-02 -> r2
      neighbors:
      - ifname: port-channel2
        node: r2
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 9
      linkindex: 10
      name: h-02 -> r1
      neighbors:
      - ifname: Ethernet4
        node: r1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 9
      linkindex: 11
      name: h-02 -> r1
      neighbors:
      - ifname: Ethernet5
        node: r1
      type: p2p
    - ifindex: 3
      ifname: eth3
      lag:
        _parentindex: 10
      linkindex: 14
      name: h-02 -> r2
      neighbors:
      - ifname: Ethernet8
        node: r2
      type: p2p
    - ifindex: 4
      ifname: eth4
      lag:
        _parentindex: 10
      linkindex: 15
      name: h-02 -> r2
      neighbors:
      - ifname: Ethernet9
        node: r2
      type: p2p
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.4/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.104
      mac: 08:4f:a9:00:00:04
    module:
    - lag
    - vlan
    name: h-02
  h2-01:
    af:
      ipv4: true
    box: debian/bookworm64
    device: frr
    id: 5
    interfaces:
    - ifindex: 30000
      ifname: bond1
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] h2-01 -> [r1,r2]'
      neighbors:
      - ifname: port-channel10
        ipv4: 172.16.0.1/24
        node: r1
      - ifname: port-channel10
        ipv4: 172.16.0.2/24
        node: r2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 1
      linkindex: 16
      name: h2-01 -> r1
      neighbors:
      - ifname: Ethernet6
        node: r1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 1
      linkindex: 17
      name: h2-01 -> r2
      neighbors:
      - ifname: Ethernet2
        node: r2
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: vlan1000
      ipv4: 172.16.0.5/24
      name: VLAN red (1000) -> [r2,r1,h2-02]
      neighbors:
      - ifname: Vlan1000
        ipv4: 172.16.0.2/24
        node: r2
      - ifname: Vlan1000
        ipv4: 172.16.0.1/24
        node: r1
      - ifname: vlan1000
        ipv4: 172.16.0.6/24
        node: h2-02
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.5/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.105
      mac: 08:4f:a9:00:00:05
    module:
    - lag
    - vlan
    name: h2-01
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  h2-02:
    af:
      ipv4: true
    box: debian/bookworm64
    device: frr
    id: 6
    interfaces:
    - ifindex: 30000
      ifname: bond1
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] h2-02 -> [r1,r2]'
      neighbors:
      - ifname: port-channel11
        ipv4: 172.16.0.1/24
        node: r1
      - ifname: port-channel11
        ipv4: 172.16.0.2/24
        node: r2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 1
      ifname: eth1
      lag:
        _parentindex: 1
      linkindex: 18
      name: h2-02 -> r1
      neighbors:
      - ifname: Ethernet7
        node: r1
      type: p2p
    - ifindex: 2
      ifname: eth2
      lag:
        _parentindex: 1
      linkindex: 19
      name: h2-02 -> r2
      neighbors:
      - ifname: Ethernet3
        node: r2
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: vlan1000
      ipv4: 172.16.0.6/24
      name: VLAN red (1000) -> [h2-01,r2,r1]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.5/24
        node: h2-01
      - ifname: Vlan1000
        ipv4: 172.16.0.2/24
        node: r2
      - ifname: Vlan1000
        ipv4: 172.16.0.1/24
        node: r1
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: lo
      ipv4: 10.0.0.6/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: eth0
      ipv4: 192.168.121.106
      mac: 08:4f:a9:00:00:06
    module:
    - lag
    - vlan
    name: h2-02
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  r1:
    af:
      ipv4: true
    box: arista/veos
    device: eos
    id: 1
    interfaces:
    - ifindex: 1
      ifname: Ethernet1
      lag:
        mlag:
          ifindex: 4094
          peer: 169.254.127.1
          peergroup: 1
          self: 169.254.127.0/31
          vlan: 4094
      linkindex: 1
      name: r1 -> r2
      neighbors:
      - ifname: Ethernet1
        node: r2
      type: p2p
    - ifindex: 30000
      ifname: port-channel8
      lag:
        ifindex: 8
        lacp: fast
        lacp_mode: active
      name: r1 -> h-01
      neighbors:
      - ifname: bond8
        node: h-01
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30001
      ifname: port-channel9
      lag:
        ifindex: 9
        lacp: fast
        lacp_mode: active
      name: r1 -> h-02
      neighbors:
      - ifname: bond9
        node: h-02
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30002
      ifname: port-channel10
      lag:
        _mlag: true
        ifindex: 10
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] r1 -> [h2-01,r2]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.5/24
        node: h2-01
      - ifname: port-channel10
        ipv4: 172.16.0.2/24
        node: r2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30003
      ifname: port-channel11
      lag:
        _mlag: true
        ifindex: 11
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] r1 -> [h2-02,r2]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.6/24
        node: h2-02
      - ifname: port-channel11
        ipv4: 172.16.0.2/24
        node: r2
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 2
      ifname: Ethernet2
      lag:
        _parentindex: 8
      linkindex: 8
      name: r1 -> h-01
      neighbors:
      - ifname: eth1
        node: h-01
      type: p2p
    - ifindex: 3
      ifname: Ethernet3
      lag:
        _parentindex: 8
      linkindex: 9
      name: r1 -> h-01
      neighbors:
      - ifname: eth2
        node: h-01
      type: p2p
    - ifindex: 4
      ifname: Ethernet4
      lag:
        _parentindex: 9
      linkindex: 10
      name: r1 -> h-02
      neighbors:
      - ifname: eth1
        node: h-02
      type: p2p
    - ifindex: 5
      ifname: Ethernet5
      lag:
        _parentindex: 9
      linkindex: 11
      name: r1 -> h-02
      neighbors:
      - ifname: eth2
        node: h-02
      type: p2p
    - ifindex: 6
      ifname: Ethernet6
      lag:
        _parentindex: 10
      linkindex: 16
      name: r1 -> h2-01
      neighbors:
      - ifname: eth1
        node: h2-01
      type: p2p
    - ifindex: 7
      ifname: Ethernet7
      lag:
        _parentindex: 11
      linkindex: 18
      name: r1 -> h2-02
      neighbors:
      - ifname: eth1
        node: h2-02
      type: p2p
    - bridge_group: 1
      ifindex: 40000
      ifname: Vlan1000
      ipv4: 172.16.0.1/24
      name: VLAN red (1000) -> [h2-01,r2,h2-02]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.5/24
        node: h2-01
      - ifname: Vlan1000
        ipv4: 172.16.0.2/24
        node: r2
      - ifname: vlan1000
        ipv4: 172.16.0.6/24
        node: h2-02
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    loopback:
      ifindex: 0
      ifname: Loopback0
      ipv4: 10.0.0.1/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: Management1
      ipv4: 192.168.121.101
      mac: 08:4f:a9:00:00:01
    module:
    - lag
    - vlan
    name: r1
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
  r2:
    _set_ifindex: true
    af:
      ipv4: true
    box: arista/veos
    device: eos
    id: 2
    interfaces:
    - ifindex: 1
      ifname: Ethernet1
      lag:
        mlag:
          ifindex: 4094
          peer: 169.254.127.0
          peergroup: 1
          self: 169.254.127.1/31
          vlan: 4094
      linkindex: 1
      name: r2 -> r1
      neighbors:
      - ifname: Ethernet1
        node: r1
      type: p2p
    - ifindex: 2
      ifname: Ethernet2
      lag:
        _parentindex: 10
      linkindex: 17
      name: r2 -> h2-01
      neighbors:
      - ifname: eth2
        node: h2-01
      type: p2p
    - ifindex: 3
      ifname: Ethernet3
      lag:
        _parentindex: 11
      linkindex: 19
      name: r2 -> h2-02
      neighbors:
      - ifname: eth2
        node: h2-02
      type: p2p
    - ifindex: 4
      ifname: Ethernet4
      linkindex: 0
      neighbors: []
      remote_id: 2
      remote_ifindex: 666
      type: p2p
    - ifindex: 5
      ifname: Ethernet5
      linkindex: 0
      neighbors: []
      remote_id: 2
      remote_ifindex: 666
      type: p2p
    - ifindex: 6
      ifname: Ethernet6
      lag:
        _parentindex: 1
      linkindex: 12
      name: r2 -> h-01
      neighbors:
      - ifname: eth3
        node: h-01
      type: p2p
    - ifindex: 7
      ifname: Ethernet7
      lag:
        _parentindex: 1
      linkindex: 13
      name: r2 -> h-01
      neighbors:
      - ifname: eth4
        node: h-01
      type: p2p
    - ifindex: 8
      ifname: Ethernet8
      lag:
        _parentindex: 2
      linkindex: 14
      name: r2 -> h-02
      neighbors:
      - ifname: eth3
        node: h-02
      type: p2p
    - ifindex: 9
      ifname: Ethernet9
      lag:
        _parentindex: 2
      linkindex: 15
      name: r2 -> h-02
      neighbors:
      - ifname: eth4
        node: h-02
      type: p2p
    - ifindex: 7
      ifname: port-channel1
      lag:
        ifindex: 1
        lacp: fast
        lacp_mode: active
      name: r2 -> h-01
      neighbors:
      - ifname: bond9
        node: h-01
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 9
      ifname: port-channel2
      lag:
        ifindex: 2
        lacp: fast
        lacp_mode: active
      name: r2 -> h-02
      neighbors:
      - ifname: bond10
        node: h-02
      pool: l2only
      type: lag
      virtual_interface: true
    - ifindex: 30000
      ifname: port-channel10
      lag:
        _mlag: true
        ifindex: 10
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] r2 -> [h2-01,r1]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.5/24
        node: h2-01
      - ifname: port-channel10
        ipv4: 172.16.0.1/24
        node: r1
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - ifindex: 30001
      ifname: port-channel11
      lag:
        _mlag: true
        ifindex: 11
        lacp: fast
        lacp_mode: active
      name: '[Access VLAN red] r2 -> [h2-02,r1]'
      neighbors:
      - ifname: bond1
        ipv4: 172.16.0.6/24
        node: h2-02
      - ifname: port-channel11
        ipv4: 172.16.0.1/24
        node: r1
      type: lag
      virtual_interface: true
      vlan:
        access: red
        access_id: 1000
    - bridge_group: 1
      ifindex: 40000
      ifname: Vlan1000
      ipv4: 172.16.0.2/24
      name: VLAN red (1000) -> [h2-01,r1,h2-02]
      neighbors:
      - ifname: vlan1000
        ipv4: 172.16.0.5/24
        node: h2-01
      - ifname: Vlan1000
        ipv4: 172.16.0.1/24
        node: r1
      - ifname: vlan1000
        ipv4: 172.16.0.6/24
        node: h2-02
      type: svi
      virtual_interface: true
      vlan:
        mode: irb
        name: red
    lag:
      lacp: fast
      lacp_mode: active
    libvirt:
      nic_adapter_count: 10
    loopback:
      ifindex: 0
      ifname: Loopback0
      ipv4: 10.0.0.2/32
      neighbors: []
      type: loopback
      virtual_interface: true
    mgmt:
      ifname: Management1
      ipv4: 192.168.121.102
      mac: 08:4f:a9:00:00:02
    module:
    - lag
    - vlan
    name: r2
    vlan:
      max_bridge_group: 1
    vlans:
      red:
        bridge_group: 1
        id: 1000
        mode: irb
        prefix:
          allocation: id_based
          ipv4: 172.16.0.0/24
plugin:
- node.clone
provider: libvirt
vlans:
  red:
    host_count: 0
    id: 1000
    neighbors:
    - ifname: vlan1000
      ipv4: 172.16.0.5/24
      node: h2-01
    - ifname: Vlan1000
      ipv4: 172.16.0.2/24
      node: r2
    - ifname: Vlan1000
      ipv4: 172.16.0.1/24
      node: r1
    - ifname: vlan1000
      ipv4: 172.16.0.6/24
      node: h2-02
    prefix:
      allocation: id_based
      ipv4: 172.16.0.0/24
